### PR TITLE
[Android] Extending Android datetime picker dialogs

### PR DESCRIPTION
**Pull request by dkarzon**
### Description of Change ###

This change makes the Android Date and Time picker renderers easier to customise by virtualising the function that creates the dialog allowing custom dialogs to be used. The most common usage of this will be to add custom buttons to the dialogs.

    // Creates a Date picker dialog with a clear button
    public class DatePickerDialogWithClear : DatePickerDialog
    {
        EventHandler<DateSetEventArgs> _callBack;

        public DatePickerDialogWithClear(Context context, EventHandler<DateSetEventArgs> callBack, int year, int monthOfYear, int dayOfMonth)
            : base(context, callBack, year, monthOfYear, dayOfMonth)
        {
            _callBack = callBack;
            SetButton((int)DialogButtonType.Neutral, "Clear", (sender, args) =>
            {
                _callBack(this, new DateSetEventArgsWithClear());
            });
        }
    }

### API Changes ###

Added:
 - protected virtual TimePickerDialog CreateTimePickerDialog

Changed:
 - void CreateDatePickerDialog -> protected virtual DatePickerDialog CreateDatePickerDialog
 - Forms.IsLollipopOrNewer is now public so the same logic can be used outside of the Xamarin.Forms project

### Behavioral Changes ###

This has no changes for existing apps but will allow developers to more easily extend the Date and Time pickers on Android.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [ ] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/Forms.cs [hunk 1]
```diff
@@ -39,7 +39,10 @@ public static class Forms
 		public static bool IsInitialized { get; private set; }
 		static bool FlagsSet { get; set; }
 
-		internal static bool IsLollipopOrNewer
+		/// <summary>
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Android/Forms.cs]: No need for this change

File: Xamarin.Forms.Platform.Android/Renderers/DatePickerRenderer.cs [hunk 2]
```diff
@@ -113,20 +113,17 @@ internal override void OnFocusChangeRequested(object sender, VisualElement.Focus
...
 		{
 			DatePicker view = Element;
-			_dialog = new DatePickerDialog(Context, (o, e) =>
+			var dialog = new DatePickerDialog(Context, (o, e) =>
 			{
 				view.Date = e.Date;
 				((IElementController)view).SetValueFromRenderer(VisualElement.IsFocusedPropertyKey, false);
 				Control.ClearFocus();
-
-				if (Forms.IsLollipopOrNewer)
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Android/Renderers/DatePickerRenderer.cs]: are we sure we don't need this now ?

File: Xamarin.Forms.Platform.Android/Forms.cs (hunk 1 above)
Comment (dkarzon) [reply Xamarin.Forms.Platform.Android/Forms.cs]: There is no reason to keep this internal. It makes it easier to use the same logic in custom renderers as the built in ones.

File: Xamarin.Forms.Platform.Android/Renderers/DatePickerRenderer.cs (hunk 2 above)
Comment (dkarzon) [reply Xamarin.Forms.Platform.Android/Renderers/DatePickerRenderer.cs]: This event in unsubscribed on focus changed and when the renderer is disposed.

File: Xamarin.Forms.Platform.Android/Forms.cs (hunk 1 above)
Comment (rmarinho) [reply Xamarin.Forms.Platform.Android/Forms.cs]: if it's public we have to maintain it and can't remove in the future, we would expose a public api but we haven't worked on it. I would prefer we keep this as is. 

Thanks

**Comment by rmarinho** [comment]
Failing tests:

Xamarin.Forms.Controls.Issues.InputTransparentTests.VerifyInputTransparent("TimePicker")_lg_nexus_5-6_0_1
Xamarin.Forms.Controls.Issues.GestureBubblingTests.VerifyTapBubbling("TimePicker",False)_lg_nexus_5-6_0_1
Xamarin.Forms.Controls.Issues.Bugzilla42074.TimePickerCancelShouldUnfocus_lg_nexus_5-6_0_1

File: Xamarin.Forms.Platform.Android/Forms.cs (hunk 1 above)
Comment (dkarzon) [reply Xamarin.Forms.Platform.Android/Forms.cs]: Reverted this change back so the property is internal 👍

**Comment by rmarinho** [comment]
Hi @dkarzon  you have 3 failing  tests on API23 and API25 can you take a look at those ?

```
Xamarin.Forms.Controls.Issues.InputTransparentTests.VerifyInputTransparent("TimePicker")
Xamarin.Forms.Controls.Issues.GestureBubblingTests.VerifyTapBubbling("TimePicker",False)
Xamarin.Forms.Controls.Issues.Bugzilla42074.TimePickerCancelShouldUnfocus
```

Thanks

**Comment by dkarzon** [comment]
@rmarinho Thanks, will take a look. 👍

**Comment by dkarzon** [comment]
@rmarinho How do I run these tests locally?

**Comment by rmarinho** [comment]
@dkarzon  https://developer.xamarin.com/guides/testcloud/uitest/working-with/testing-on-devices/

**Comment by rmarinho** [comment]
any luck with this @dkarzon  ?!
Thanks

**Comment by dkarzon** [comment]
@rmarinho I haven't had a chance to look at it just yet. Hopefully soon.

**Comment by hartez** [comment]
@rmarinho I just ran the failing tests locally and they all passed. I'm guessing this is a CI hiccup - I'm queueing up another run.

**Comment by dkarzon** [comment]
Ahhhh I did not want to do that :/
Will close this one and reopen it properly.

**Comment by dkarzon** [comment]
Reopened as #2201

**Pull request closed without merging**